### DISCUSSION TITLE
Black Textures on iOS

### DISCUSSION THREAD

**Discussion by Aubi0ne** [google/model-viewer#2565] [Q&A]
Hello, I am having an issue with two gltf models on iOS showing black texture instead of the real ones. This only happens on the 3D viewport on iOS (safari and chrome) but it still displays well in ar. I have also tested on 3d viewport on pc and android as well as ar on android and it works fine. Any idea to fix it ?

iOS Safari
![199259142_3001070990220869_7608761589287570170_n](https://user-images.githubusercontent.com/60987844/124973056-5b01eb00-e02b-11eb-8000-94736812cb4f.jpg)

Android Chrome
![211330342_595343521454873_5488914306281258576_n](https://user-images.githubusercontent.com/60987844/124973267-98ff0f00-e02b-11eb-8f2f-df15f3e12ee9.jpg)

[Link to the model](https://ar.linkerplace.com/AR/Bochner_Its_Always_Something%20_%20It_Goes_Without_Saying?type=wall)

**Comment by elalish** [comment]
Thanks, I can repro. Is there anything unusual about your model? It looks like the lighting is working, but just not the texture. Have you tried packing it as a GLB?

**Reply by Aubi0ne** [reply]
No, I haven't tried the glb format yet because I don't have access to my test iPhone for the weekend but I can provide you with a link to the model. The models I use have been provided to me, this one among others, and I only know a few basics of 3D and at my level, this model seems to me quite identical to the others. 
[Link to download the models](https://we.tl/t-dskttwGAPM)

**Reply by Aubi0ne** [reply]
Ok it seems that reducing the size of the texture fix the problem. Not sure to understand why but it works now :)

**Reply by ferb7nft** [reply]
What size did you reduce the texture to? Cause a 1000px by 500px texture image wrapped around the object is still black on iOS for me.

**Reply by elalish** [reply]
Did you try a power-of-2 texture? I didn't think that was an iOS requirement, but it crops up from time to time...

**Reply by ferb7nft** [reply]
I will try. But I just tested this .usdz from the nasa website, and even it had a black texture on iOS. I tried increasing the exposure in the model-viewer, but that made the item shiny but still black (so that is not the problem). Here is the file that doesn't work on iOS, but does on desktop: 
[Earth_1_12756.usdz.zip](https://github.com/google/model-viewer/files/6851050/Earth_1_12756.usdz.zip)

**Reply by ferb7nft** [reply]
@elalish just tried 2048px by 1024px image texture and the object is still black on iOS, but fine on Desktop. The texture is also a PNG wrap, does that matter? Should it be JPG? Does USDZ not support image textures? I export the model from blender as a .dae, and open that in Xcode, and then export it as .usdz.

**Reply by ferb7nft** [reply]
@elalish I just tried a 512px by 512px texture (small, 1:1, power-of-two dimensions) but still have the black texture. Shows up on iOS and iPadOS. Any ideas?

**Reply by elalish** [reply]
Can you post a repro link?

**Reply by ferb7nft** [reply]
Ok just testing no ios-src, only .glb src with the square texture, seeing if it converts. Then I will share a link if it doesn't work.

**Reply by ferb7nft** [reply]
@elalish Here is the reproducible error website: https://ipfs.io/ipfs/QmQhCfU4xq7siuR6owefi6QYQJtoZSuX28yvcKBvDJwx9E/

And here are the 3d files and index.html code: 
[resources.zip](https://github.com/google/model-viewer/files/6857055/resources.zip)

**Reply by elalish** [reply]
Here you're supplying your own ios-src, so presumably the error is in whatever software created that. Does it open on a mac?

**Reply by ferb7nft** [reply]
The error is in the ios-src file! I just used only src glb, no iOS-src specified, and the model viewer converter successfully created a usdz for view on iOS and iPadOS with no black texture! 

Strangely, the broken ios-src files open just fine on macOS Xcode and macOS Safari with proper textures. And the original usdz was downloaded from the nasa website and didn't work on iOS. Furthermore, I manually added the texture in Blender and Xcode to produce the usdz which also didn't work as an ios-src. So, what is best practice to produce working iOS-src files, so the website user doesn't need to wait for the glb conversion? @elalish

**Reply by elalish** [reply]
The GLB conversion is dramatically faster than downloading an extra USDZ file, so I'd recommend going the auto-conversion route. If you want the same code we use for auto-conversion, you can try https://threejs.org/editor/

**Reply by ferb7nft** [reply]
> The GLB conversion is dramatically faster than downloading an extra USDZ file, so I'd recommend going the auto-conversion route. If you want the same code we use for auto-conversion, you can try https://threejs.org/editor/

Thank you so much! I will go the auto-convert route then
